mediaapi/routing/download.go

```diff
@@ -767,19 +823,68 @@ func (r *downloadRequest) fetchRemoteFile(
+	resp, err := r.fedClient.DownloadMedia(ctx, r.origin, r.MediaMetadata.Origin, string(r.MediaMetadata.MediaID))
 	if err != nil || (resp != nil && resp.StatusCode != http.StatusOK) {
-		if resp != nil && resp.StatusCode == http.StatusNotFound {
-			return "", false, fmt.Errorf("File with media ID %q does not exist on %s", r.MediaMetadata.MediaID, r.MediaMetadata.Origin)
+		isMultiPart = false
+		// try again on the unauthed endpoint
+		// create request for remote file
+		resp, err = client.CreateMediaDownloadRequest(ctx, r.MediaMetadata.Origin, string(r.MediaMetadata.MediaID))
+		if err != nil || (resp != nil && resp.StatusCode != http.StatusOK) {
+			if resp != nil && resp.StatusCode == http.StatusNotFound {
+				return "", false, fmt.Errorf("File with media ID %q does not exist on %s", r.MediaMetadata.MediaID, r.MediaMetadata.Origin)
+			}
+			return "", false, fmt.Errorf("file with media ID %q could not be downloaded from %s", r.MediaMetadata.MediaID, r.MediaMetadata.Origin)
 		}
-		return "", false, fmt.Errorf("file with media ID %q could not be downloaded from %s", r.MediaMetadata.MediaID, r.MediaMetadata.Origin)
 	}
 	defer resp.Body.Close() // nolint: errcheck
 
-	// The reader returned here will be limited either by the Content-Length
-	// and/or the configured maximum media size.
-	contentLength, reader, parseErr := r.GetContentLengthAndReader(resp.Header.Get("Content-Length"), &resp.Body, maxFileSizeBytes)
+	// If this wasn't a multipart response, set the Content-Type now. Will be overwritten
+	// by the multipart Content-Type below.
+	r.MediaMetadata.ContentType = types.ContentType(resp.Header.Get("Content-Type"))
+
+	var contentLength int64
+	var reader io.Reader
+	var parseErr error
+	if isMultiPart {
+		r.Logger.Debug("Downloaded file using authenticated endpoint")
+		var params map[string]string
+		_, params, err = mime.ParseMediaType(resp.Header.Get("Content-Type"))
+		if err != nil {
+			panic(err)
+		}
+		if params["boundary"] == "" {
+			return "", false, fmt.Errorf("no boundary header found on media %s from %s", r.MediaMetadata.MediaID, r.MediaMetadata.Origin)
+		}
+		mr := multipart.NewReader(resp.Body, params["boundary"])
+
+		// Get the first, JSON, part
+		p, multipartErr := mr.NextPart()
+		if multipartErr != nil {
+			return "", false, multipartErr
+		}
+
+		if p.Header.Get("Content-Type") != "application/json" {
+			return "", false, fmt.Errorf("first part of the response must be application/json")
+		}
+		// TODO: Once something is defined, parse the JSON content
+
+		// Get the actual media content
+		p, multipartErr = mr.NextPart()
+		if multipartErr != nil {
+			return "", false, multipartErr
+		}
+
+		contentLength, reader, parseErr = r.GetContentLengthAndReader(p.Header.Get("Content-Length"), p, maxFileSizeBytes)
+		// For multipart requests, we need to get the Content-Type of the second part, which is the actual media
+		r.MediaMetadata.ContentType = types.ContentType(p.Header.Get("Content-Type"))
```

Comment:
I think you're missing redirect handling here? https://spec.matrix.org/v1.11/server-server-api/#get_matrixfederationv1mediadownloadmediaid

> The second part is either:
> 1. [...]
> 2. or a `Location` header to redirect the caller to where the media can be retrieved. The URL at `Location` SHOULD have appropriate `Content-Type` and `Content-Disposition` headers which describe the media.
>
>    When `Location` is present, servers SHOULD NOT cache the URL. The remote server may have applied time limits on its validity. If the caller requires an up-to-date URL, it SHOULD re-request the media download.

Reply:
Thanks for the fix! :)